packages/app/src/cli/services/function/replay.ts

```diff
@@ -92,21 +92,36 @@ async function getFunctionRunData(functionRunsDir: string, functionHandle: strin
       return splitFilename[3] === 'extensions' && splitFilename[4] === functionHandle
     })
     .reverse()
-  const latestFunctionRunFileNames = allFunctionRunFileNames.slice(0, LOG_SELECTOR_LIMIT)
-  const functionRunFilePaths = latestFunctionRunFileNames.map((functionRunFile) =>
-    joinPath(functionRunsDir, functionRunFile),
-  )
 
-  const functionRunData = await Promise.all(
-    functionRunFilePaths.map(async (functionRunFilePath) => {
-      const fileData = await readFile(functionRunFilePath)
-      const parsedData = JSON.parse(fileData)
-      return {
-        ...parsedData,
-        identifier: getIdentifierFromFilename(functionRunFilePath),
-      }
-    }),
-  )
+  let functionRunData = []
+  for (let i = 0; i < allFunctionRunFileNames.length; i += LOG_SELECTOR_LIMIT) {
+    const currentFunctionRunFileNameChunk = allFunctionRunFileNames.slice(i, i + LOG_SELECTOR_LIMIT)
+    const functionRunFilePaths = currentFunctionRunFileNameChunk.map((functionRunFile) =>
+      joinPath(functionRunsDir, functionRunFile),
+    )
+
+    // eslint-disable-next-line no-await-in-loop
+    const functionRunDataFromChunk = await Promise.all(
+      functionRunFilePaths.map(async (functionRunFilePath) => {
+        const fileData = await readFile(functionRunFilePath)
+        const parsedData = JSON.parse(fileData)
+        return {
+          ...parsedData,
+          identifier: getIdentifierFromFilename(functionRunFilePath),
+        }
+      }),
+    )
+
+    functionRunData = functionRunData.concat(
+      functionRunDataFromChunk.filter((run) => {
+        return run.payload.input !== null
+      }),
+    )
+
+    if (functionRunData.length >= LOG_SELECTOR_LIMIT) {
+      break
+    }
+  }
```

Comment:
If we naively filtered the list, then it's possible for users to get empty or small lists in the event that their first 100 logs have no input. Since I still don't want to read the entire directory at once, I've set up an algorithm here that reads 100 files at a time, filtering out those without input and halting once we have at least 100 included logs.

In the likely event that the user has no events with no input, then this loop will only iterate once. In the event that the user has any missing inputs, it will likely iterate twice. In the rare case that we need to look at more then 200 logs to find 100 valid logs, it will eventually parse the entire directory searching for valid logs.

Reply:
Alternatives:
- Read the entire directory, filter out empty inputs, take the first 100
  - I've avoided this approach since reading the entire directory could be a slow operation depending on the directories size
- Ignore the issue and accept that replay lists could be small or empty in the event that a significant portion of the last 100 events have no input
- Indicate if a log is 'replay'able in the filename, so that we can filter them out at the same time as we filter out extension handles